feat(opennav_docking): add on-demand detector client & unit tests for dock plugins (#5015)

bkoensgen opened the pull request:
* Add detector client initialization in **SimpleNonChargingDock**
* Mirror the same logic in **SimpleChargingDock**
* Expand unit-test coverage for both dock plugins

<!-- Please fill out the following pull-request template for non-trivial changes to help us process your PR faster and more efficiently. -->

---

## Basic Info

| Info | Value |
|------|-------|
| Ticket(s) this addresses | #5015 |
| Primary OS tested on | Official Nav2 **Rolling** Docker image |
| Robotic platform tested on | Unit tests only (`colcon test`) |
| Does this PR contain AI-generated software? | No |

---

## Description of contribution in a few bullet points

* Introduce **on-demand external-detector control** for both charging and non-charging dock plugins  
  * New parameters: `detector_service_name`, `detector_service_timeout`, `subscribe_toggle`
  * Detector started/stopped via `std_srvs/Trigger` only when first needed
* **SimpleNonChargingDock** now creates the detector client (symmetry with charging dock)
* **Unit tests** added to validate service call logic and TF handling

---

## Description of documentation updates required

* Added the three new parameters to `nav2_docking/README.md`  

---

## Description of how this change was tested

* `colcon test --packages-select opennav_docking` → **all tests pass**
* Manual sanity check: plugins load without warnings inside the Rolling dev-container

---

#### For Maintainers <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in **docs.nav2.org**
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins are added to the plugins page
- [ ] If BT Node, additionally: add to BT's XML index of nodes for Groot, BT package's readme table, and BT library lists

SteveMacenski commented on nav2_docking/opennav_docking/include/opennav_docking/simple_charging_dock.hpp:
```diff
@@ -51,19 +52,23 @@ class SimpleChargingDock : public opennav_docking_core::ChargingDock
 
   /**
-   * @brief Method to cleanup resources used on shutdown.
+   * @brief Final cleanup, in case deactivate() was skipped.
```
In general, remove all comment changes to these headers, they add unnecessary and sometimes incorrect information.

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,7 +342,85 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+
+void SimpleChargingDock::cleanup()
+{
+  stopDetection();
```
Remove, this should be already stopped by `deactivate`.

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,7 +342,85 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+
+  // 1. Service STOP request
+  //    Fire-and-forget; detector should teardown itself
+  if (detector_client_ && detector_client_->service_is_ready()) {
```
With the Nav2 Utils Service Client, you can just do a `detector_client_ && detector_client_->invoke()`. Make sure to use the API with the timeout, we cannot have indefinite spins. There has to be error handling somewhere on failures.

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,7 +342,85 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+  //    Fire-and-forget; detector should teardown itself
+  if (detector_client_ && detector_client_->service_is_ready()) {
+    auto req = std::make_shared<std_srvs::srv::Trigger::Request>();
+    detector_client_->async_send_request(req);
```
For example: this is never spun to actually know if this was completed.

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -133,11 +154,6 @@ void SimpleChargingDock::configure(
 
   if (use_external_detection_pose_) {
     dock_pose_.header.stamp = rclcpp::Time(0);
-    dock_pose_sub_ = node_->create_subscription<geometry_msgs::msg::PoseStamped>(
```
If `!subscribe_toggle_`, I think we should create this, no?

SteveMacenski replied:
I think the code should move back up here. You also don't need to check `!detected_pose_sub_`

SteveMacenski replied:
@bkoensgen this was not completed :-)

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,7 +342,85 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+                  detector_service_name_.c_str(), detector_service_timeout_);
+    } else {
+      auto req = std::make_shared<std_srvs::srv::Trigger::Request>();
+      auto future = detector_client_->async_send_request(req);
```
`invoke` which does the `spin` this is missing

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,7 +342,85 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+      });
+  }
+
+  detector_state_ = DetectorState::WAITING_START;
```
This state is never queried, so is it useful?

SteveMacenski reviewed (changes requested):
Good first stab! I have some questions and suggestions, but largely its a good starting point for a couple of hopefully quick iterations. The changes requested in this plugin are the same in both, but I only reviewed the one just to avoid duplicating every comment. Take each comment for both plugin files

bkoensgen commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -196,6 +212,10 @@ bool SimpleChargingDock::getRefinedPose(geometry_msgs::msg::PoseStamped & pose,
   }
 
+  if (detector_state_ == DetectorState::OFF) {
+    startDetection();
```
hi @SteveMacenski,

Thanks for the question,

You're right that having the DockingServer explicitly control when detection starts/stops would be more optimal than the current approach where plugins manage it internally via getRefinedPose().

The current PR implements the detector control within plugins (started by getRefinedPose(), stopped by deactivate()). To implement your suggested architecture, we'd need to:

1. Add public API methods (e.g., `startDetectionProcess()`/`stopDetectionProcess()`) to the dock plugin interface
2. Modify DockingServer to call these at appropriate stages (after staging pose, after docking complete)
3. Update getRefinedPose() to assume detection is already active

?: Should I expand this PR to include these DockingServer-level changes, or would you prefer to merge the current plugin-internal implementation (after fixing the ServiceClient issues) and tackle server-side control in a follow-up PR?

Happy to go either direction ;just want to make sure we align on scope

SteveMacenski commented:
@bkoensgen re-request a review from me in the github UX when you're ready for me to take a look again (and CI is building :wink: )

SteveMacenski commented:
@bkoensgen you may want to compile locally and test rather than using CI as your builder. you'd probably have faster iteration cycles.

bkoensgen commented:
@SteveMacenski

Thanks for the detailed review and guidance! I've pushed up the changes addressing your feedback.

The biggest change is that I've moved the detector lifecycle management (`start/stopDetectionProcess`) into the plugin API itself. The `DockingServer` now just calls these methods, which gives individual plugins full control over how and when they manage their perception resources (like starting/stopping a GPU-intensive node). This felt a lot cleaner and more robust.

Here’s a quick rundown of the other changes:

*   **Service Client & Dependencies:** Switched over to `nav2_util::ServiceClient` as you suggested and cleaned up the CMakeLists to properly link `std_srvs`.
*   **Test Coverage:** Beefed up the unit tests for both `SimpleChargingDock` and `SimpleNonChargingDock`. They now cover the new `subscribe_toggle` logic and the service failure edge cases we discussed.
*   **Test Reliability:** That also gave me a chance to fix a race condition in the Python smoke tests related to TF propagation. They should be much more stable in CI now.
*   **Git History:** Squashed everything down to a single commit for a clean merge.

Should be in good shape for another look now. Let me know what you think!

SteveMacenski commented on nav2_docking/opennav_docking_core/include/opennav_docking_core/non_charging_dock.hpp:
```diff
@@ -125,6 +125,16 @@ class NonChargingDock : public ChargingDock
+  /**
+   * @brief Start any detection pipelines required for pose refinement.
+   */
+  virtual void startDetectionProcess() = 0;
```
I think both of these can be initialized as empty methods, ie `{}` so that they are not required to be implemented if there is no enable/disable logic. I'd add into the doxygen that this can be implemented if the detector is able to be toggled on and off. Otherwise, do not implement.

SteveMacenski commented on nav2_docking/opennav_docking_core/include/opennav_docking_core/charging_dock.hpp:
```diff
@@ -124,6 +124,16 @@ class ChargingDock
+  /**
+   * @brief Start any detection pipelines required for pose refinement.
+   */
+  virtual void startDetectionProcess() = 0;
```
Ditto

SteveMacenski commented on nav2_docking/opennav_docking/test/test_docking_server.py:
```diff
@@ -127,8 +129,27 @@ def setUp(self) -> None:
+                self.node.get_logger().info(f'Node {node_name} is active.')
+                return
+            time.sleep(0.5)
+        self.fail(f'Node {node_name} did not become active within {timeout_sec} seconds.')
```
This is a log, not actually returning a failure or throwing an exception. I think this should be propagated to fail the test, else we wouldn't know

bkoensgen replied:
Thanks for raising this

Quick confirmation before I call it a night: `self.fail()` raises `AssertionError` unconditionally (see std-lib docs), so the test is already marked **FAILED** when the node stays inactive. I’ll add a small `# raises AssertionError` comment in my next push to make that explicit.

¹ https://docs.python.org/3/library/unittest.html#unittest.TestCase.fail

I still have a couple of test tweaks to finish; I’ll batch-reply to the remaining threads and push the updated diff tomorrow.

Let me know if I missed anything in the meantime

SteveMacenski replied:
Got it - this is right then. We have other wrappers were things like `self.fail` and `self.info` map to logging. I didn't check that this had a unittesting base class.

SteveMacenski commented on nav2_docking/README.md:
```diff
@@ -36,12 +36,14 @@ The `ChargingDock` and `NonChargingDock` plugins are the heart of the customizab
+5. Enter a vision-control loop where the robot attempts to reach the docking pose while it's actively being refined by the vision system.
+6. Exit the vision-control loop once contact has been detected or charging has started (if applicable).
+7. Wait until charging starts (if applicable) and return success.
+8. Call the dock's plugin `stopDetectionProcess()` method to deactivate any external detection mechanisms (this is also called on failure/cancellation).
```
```suggestion
8. Call the dock's plugin `stopDetectionProcess()` method to deactivate any external detection mechanisms.
```

SteveMacenski replied:
TMI

SteveMacenski commented on nav2_docking/README.md:
```diff
@@ -36,12 +36,14 @@ The `ChargingDock` and `NonChargingDock` plugins are the heart of the customizab
+8. Call the dock's plugin `stopDetectionProcess()` method to deactivate any external detection mechanisms (this is also called on failure/cancellation).
 
-If anywhere this procedure is unsuccessful, `N` retries may be made by driving back to the dock's staging pose and trying again. If still unsuccessful, it will return a failure code to indicate what kind of failure occurred to the client.
+If anywhere this procedure is unsuccessful (before step 8), `N` retries may be made. This involves calling `stopDetectionProcess()`, driving back to the dock's staging pose, and then restarting the process from step 3. If still unsuccessful after retries, it will return a failure code to indicate what kind of failure occurred to the client.
```
```suggestion
If anywhere this procedure is unsuccessful (before step 8), `N` retries may be made, driving back to the dock's staging pose, and then restarting the process from step 3. If still unsuccessful after retries, it will return a failure code to indicate what kind of failure occurred to the client.
```

SteveMacenski replied:
TMI

SteveMacenski commented on nav2_docking/README.md:
```diff
@@ -151,6 +153,10 @@ There are two functions used during dock approach:
    for charging to start by calling `isCharging`.
    * This may be implemented using the `sensor_msgs/BatteryState` message to check the power status or for charging current.
    * Used for charging-typed plugins.
+   + * `startDetectionProcess`: Start any detection pipelines required for pose refinement.
```
These bullets are needing 1 less set of indentations, currently they're nested under `isCharging`. Also, then the `two functions` should be 4. 

Personally I Think these can be removed, these are just describing the 'main' functions, there are more than just these and I don't think these count as 'main' functions.

SteveMacenski commented on nav2_docking/README.md:
```diff
@@ -175,8 +181,9 @@ some robots.
 
 `getRefinedPose` can be used in two ways.
-1. A blind approach where the returned dock pose will simply be equal to whatever was passed in from the dock database. This may work with a reduced success rate on a real robot (due to global localization error), but is useful for initial testing and simulation.
+1. A blind approach where the returned dock pose will simply be equal to whatever was passed in from the dock database. This may work with a reduced success rate on a real robot (due to global localization error), but is useful for initial testing and simulation. The `start/stopDetectionProcess` methods would typically do nothing in this case.
```
Remove

SteveMacenski commented on nav2_docking/README.md:
```diff
@@ -175,8 +181,9 @@ some robots.
-1. A blind approach where the returned dock pose will simply be equal to whatever was passed in from the dock database. This may work with a reduced success rate on a real robot (due to global localization error), but is useful for initial testing and simulation.
+1. A blind approach where the returned dock pose will simply be equal to whatever was passed in from the dock database. This may work with a reduced success rate on a real robot (due to global localization error), but is useful for initial testing and simulation. The `start/stopDetectionProcess` methods would typically do nothing in this case.
 2. The more realistic use case is to use an AR marker, dock pose detection algorithm, etc. The plugin will subscribe to a `geometry_msgs/PoseStamped` topic `detected_dock_pose`. This can be used with the `image_proc/TrackMarkerNode` for Apriltags or other custom detectors for your dock. It is unlikely the detected pose is actually the pose you want to dock with, so several parameters are supplied ...
+The subscription to `detected_dock_pose` can be managed dynamically (see `subscribe_toggle` parameter). Additionally, an external detector process can be triggered via a service call (see `detector_service_name`). The `DockingServer` calls `startDetectionProcess()` before `getRefinedPose` (which is called in a loop) and `stopDetectionProcess()` after the docking action concludes (successfully or  ...
```
```suggestion
The subscription to `detected_dock_pose` can be managed dynamically (see `subscribe_toggle` parameter). Additionally, an external detector process can be triggered via a service call (see `detector_service_name`). The `DockingServer` calls `startDetectionProcess()` before `getRefinedPose` and `stopDetectionProcess()` after the docking action concludes (successfully or not). This allows plugins to manage resources like GPU usage by only running detection when needed.
```

SteveMacenski commented on nav2_docking/README.md:
```diff
@@ -257,6 +264,9 @@ Note: `dock_plugins` and either `docks` or `dock_database` are required.
 | rotate_to_dock        | Enables backward docking without requiring a sensor for detection during the final approach. When enabled, the robot approaches the staging pose facing forward with sensor coverage for dock detection; after detection, it rotates and backs into the dock using only the initially detected pose for dead reckoning. | bool |  false      |
+| detector_service_name      | Trigger service name to start/stop the detector (e.g., a camera node or an AprilTag detector node). Leave empty to disable service calls. | string | ""    |
+| detector_service_timeout   | Timeout (seconds) to wait for the detector service (if `detector_service_name` is set) to become available or respond. | double | 5.0   |
+| subscribe_toggle           | If true, dynamically subscribe/unsubscribe to `detected_dock_pose` topic when `start/stopDetectionProcess` are called. If false, the subscription is kept alive throughout the plugin's lifecycle if `use_external_detection_pose` is true. This can be useful if the detector node is always running and publishing. | bool   | true  |
```
```suggestion
| subscribe_toggle           | If true, dynamically subscribe/unsubscribe to `detected_dock_pose` topic when `start/stopDetectionProcess` are called. If false, the subscription is kept alive throughout the plugin's lifecycle if `use_external_detection_pose` is true. This can be useful if the detector node only computes values when it has a subscription. | bool   | true  |
```

SteveMacenski commented on nav2_docking/opennav_docking/include/opennav_docking/simple_charging_dock.hpp:
```diff
@@ -103,6 +105,16 @@ class SimpleChargingDock : public opennav_docking_core::ChargingDock
+  /**
+   * @brief Start external detection process (service call + subscribe).
+   */
+  void startDetectionProcess() override {startDetection();}
```
Just implement startDetectionProcess with startDetection's content, there's not a need for a new method.

Ditto with the Stop functions

SteveMacenski replied:
@bkoensgen this was not completed :-)

SteveMacenski replied:
@bkoensgen this was still not completed

SteveMacenski commented on nav2_docking/opennav_docking/include/opennav_docking/simple_charging_dock.hpp:
```diff
@@ -149,6 +161,29 @@ class SimpleChargingDock : public opennav_docking_core::ChargingDock
+  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr detected_pose_sub_;
+
+  // Simple finite-state machine for detector status
+  enum class DetectorState { OFF, ON };
```
If its just 2 states, this can simply be a bool `detector_enabled_`

SteveMacenski replied:
@bkoensgen this was not completed :-)

SteveMacenski commented on nav2_docking/opennav_docking/include/opennav_docking/simple_charging_dock.hpp:
```diff
@@ -149,6 +161,29 @@ class SimpleChargingDock : public opennav_docking_core::ChargingDock
+  DetectorState detector_state_{DetectorState::OFF};
+
+  // Internally enable detector (service + subscribe)
+  void startDetection();
```
Remove these 2 methods

SteveMacenski replied:
@bkoensgen this was not completed :-)

SteveMacenski replied:
@bkoensgen this is related to avoid

SteveMacenski commented on nav2_docking/opennav_docking/src/docking_server.cpp:
```diff
@@ -164,7 +164,10 @@ DockingServer::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
   docking_action_server_.reset();
   undocking_action_server_.reset();
-  dock_db_.reset();
+  if (dock_db_) {
```
Cleanup is the opposite of configure. This would never be called if configure isn't called, so we know `dock_db_` must exist at this point

SteveMacenski replied:
This is already handled 

```
  if (!dock_db_->initialize(node, tf2_buffer_)) {
    on_cleanup(state);
    return nav2_util::CallbackReturn::FAILURE;
  }
```

Anything that can throw is already protected

SteveMacenski commented on nav2_docking/opennav_docking/src/docking_server.cpp:
```diff
@@ -260,6 +263,9 @@ void DockingServer::dockRobot()
         get_logger(), "Robot is already docked and/or charging (if applicable), no need to dock");
       result->success = true;
       docking_action_server_->succeeded_current(result);
+      if (dock) {
```
We know dock exists, we call it in conditions above

SteveMacenski replied:
@bkoensgen

SteveMacenski commented on nav2_docking/opennav_docking/src/docking_server.cpp:
```diff
@@ -329,13 +335,15 @@ void DockingServer::dockRobot()
             stashDockData(goal->use_dock_id, dock, true);
             publishZeroVelocity();
             docking_action_server_->succeeded_current(result);
+            dock->plugin->stopDetectionProcess();
```
Move this before `succeeded_current` line. More about consistency across the codebase than a technical issue. We tend to do the very last thing as succeeding / terminating actions so that timers that can be inserted for metrics

SteveMacenski commented on nav2_docking/opennav_docking/src/docking_server.cpp:
```diff
@@ -351,6 +359,7 @@ void DockingServer::dockRobot()
         // Cancelled, preempted, or shutting down
         stashDockData(goal->use_dock_id, dock, false);
         publishZeroVelocity();
+        dock->plugin->stopDetectionProcess();
```
Should we reset here? We're going to go back around to try again, I don't think there's any reason to stop detection

bkoensgen replied:
Ah, good catch. Yeah, we definitely don't want to stop detection if we're just going to retry.

I double-checked, and this specific `if` block is for when the whole action gets cancelled or preempted. It's a hard exit, so it's our last chance to call `stopDetectionProcess()` to prevent a leak.

The actual retry logic in the `catch` block correctly leaves detection running. So I think we're good here. Let me know if you see it differently!

SteveMacenski commented on nav2_docking/opennav_docking/src/docking_server.cpp:
```diff
@@ -395,6 +404,9 @@ void DockingServer::dockRobot()
   }
 
   // Store dock state for later undocking and delete temp dock, if applicable
+  if (dock) {
```
we know `dock` exists, we don't need to  check it

SteveMacenski replied:
@bkoensgen

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -37,6 +40,14 @@ void SimpleChargingDock::configure(
+  nav2_util::declare_parameter_if_not_declared(
+    node_, name + ".detector_service_timeout", rclcpp::ParameterValue(5.0));
+  nav2_util::declare_parameter_if_not_declared(
+    node_, name + ".subscribe_toggle", rclcpp::ParameterValue(true));
```
default to false to keep current behavior

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -133,11 +155,6 @@ void SimpleChargingDock::configure(
 
   if (use_external_detection_pose_) {
```
```suggestion
  if (use_external_detection_pose_ && !subscribe_toggle_) {
```

No?

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -157,6 +174,17 @@ void SimpleChargingDock::configure(
+  // subscription alive by creating it here during configuration
+  if (!subscribe_toggle_ && use_external_detection_pose_ && !detected_pose_sub_) {
+    detected_pose_sub_ = node_->create_subscription<geometry_msgs::msg::PoseStamped>(
+      "detected_dock_pose", rclcpp::SensorDataQoS(),
```
Revert the QoS when you move it back

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +350,88 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+
+void SimpleChargingDock::deactivate()
+{
+  RCLCPP_DEBUG(node_->get_logger(), "SimpleChargingDock deactivated");
```
If you make deactivated call `stopDetection`, you can remove the explicit `stopDetection` call in the dock database. Its probably good that these reset its own resources internally.

SteveMacenski replied:
@bkoensgen

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +350,88 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+
+void SimpleChargingDock::activate()
+{
+  RCLCPP_DEBUG(node_->get_logger(), "SimpleChargingDock activated");
```
This is unused. It can be reverted into a non-override / `{}`

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +350,88 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+void SimpleChargingDock::startDetection()
+{
+  // Skip if already starting or ON
+  if (detector_state_ != DetectorState::OFF) {return;}
```
```suggestion
  if (detector_state_ == DetectorState::ON) {return;}
```

This seems more rational (though I think you're going to change this to a bool anyway)

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +350,88 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+  if (detector_client_) {
+    auto req = std::make_shared<std_srvs::srv::Trigger::Request>();
+    try {
+      detector_client_->invoke(
```
You should capture the return and make sure the success is `true`

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +350,88 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+    } catch (const std::exception & e) {
+      RCLCPP_WARN(node_->get_logger(),
+        "Detector service '%s' failed: %s",
+        detector_service_name_.c_str(), e.what());
```
This should be an error message

Also, I think the signature of the start/stop should be `bool` so that if this happens, we can return false and the application can handle that error in the docking server if it cannot be enabled by throwing an exception `opennav_docking_core::FailedToDetectDock`

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +350,88 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+  //    Only subscribe once; will set state to ON on first message
+  if (subscribe_toggle_ && !detected_pose_sub_) {
+    detected_pose_sub_ = node_->create_subscription<geometry_msgs::msg::PoseStamped>(
+      "detected_dock_pose", rclcpp::SensorDataQoS(),
```
QoS

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +350,88 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+    detected_pose_sub_ = node_->create_subscription<geometry_msgs::msg::PoseStamped>(
+      "detected_dock_pose", rclcpp::SensorDataQoS(),
+      [this](const geometry_msgs::msg::PoseStamped::SharedPtr pose) {
+        detected_dock_pose_ = *pose;
```
Should uses of `detected_dock_pose_` make sure that `detector_state_` is on? Throw otherwise

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +350,88 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+      "detected_dock_pose", rclcpp::SensorDataQoS(),
+      [this](const geometry_msgs::msg::PoseStamped::SharedPtr pose) {
+        detected_dock_pose_ = *pose;
+        detector_state_ = DetectorState::ON;
```
Remove. This is always set anyway on L384 below when we exist the function. It has no impact in the callback (ditto for the subscriber above).

SteveMacenski replied:
This is still a problem. detector_state_ is the state of the detector running, not whether or not you've received data. 

You need to update this so `detector_state_` is `true` on `startDetection` so that the state is properly managed from that perspective. Then, you can have another variable for `detected_pose_received` that can be set to `true` when we receive a message (set to false in whatever initialization / reset conditions required) that can be used in the `getRefinedPose` function.

However, I would argue that you could **probably** remove the `detector_state_` variable in that case. Right now, that's only protecting irraneous calls to `stopDetection` when stopped already and `startDetection` when already started. Given the logic of the server is otherwise sound, I don't think that is actually necessary.

You could just change the name of the variable to `detected_pose_received_`, have it set to `true` in the subscription callback, and have it initialized to `false` on `configure`, `stopDetection`, etc.

SteveMacenski commented on nav2_docking/opennav_docking/include/opennav_docking/simple_charging_dock.hpp:
```diff
@@ -149,6 +161,29 @@ class SimpleChargingDock : public opennav_docking_core::ChargingDock
+  // Detector control parameters
+  std::string detector_service_name_;
+  double detector_service_timeout_{5.0};
+  bool subscribe_toggle_{true};
```
```suggestion
  bool subscribe_toggle_{false};
```

SteveMacenski commented on nav2_docking/opennav_docking/include/opennav_docking/simple_non_charging_dock.hpp:
```diff
@@ -127,6 +139,29 @@ class SimpleNonChargingDock : public opennav_docking_core::NonChargingDock
+  // Detector control parameters
+  std::string detector_service_name_;
+  double detector_service_timeout_{5.0};
+  bool subscribe_toggle_{true};
```
```suggestion
  bool subscribe_toggle_{false};
```

SteveMacenski commented on nav2_docking/opennav_docking/src/docking_server.cpp:
```diff
@@ -326,6 +329,7 @@ void DockingServer::dockRobot()
             }
             result->success = true;
             result->num_retries = num_retries_;
+            dock->plugin->stopDetectionProcess();
```
Move after publishZeroVelocity() so the order is the same in all terminal conditions

SteveMacenski commented on nav2_docking/opennav_docking/include/opennav_docking/simple_charging_dock.hpp:
```diff
@@ -149,6 +161,28 @@ class SimpleChargingDock : public opennav_docking_core::ChargingDock
+  bool subscribe_toggle_{false};
+
+  // Client used to call the Trigger service
+  nav2_util::ServiceClient<
```
Rebase  / merge in changes. This class is now different (nav2::ServiceClient) and should be gotten from the `nav2::LifecycleNode`'s `create_client`. See https://docs.nav2.org/migration/Kilted.html#new-nav2-ros-common-nav2-lifecycle-node

SteveMacenski commented on nav2_docking/opennav_docking/include/opennav_docking/simple_charging_dock.hpp:
```diff
@@ -149,6 +161,28 @@ class SimpleChargingDock : public opennav_docking_core::ChargingDock
+    rclcpp_lifecycle::LifecycleNode::SharedPtr>::SharedPtr detector_client_;
+
+  // Dynamic subscription pointer (can be reset to release CPU/GPU)
+  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr detected_pose_sub_;
```
Ditto, see https://docs.nav2.org/migration/Kilted.html#new-nav2-ros-common-nav2-lifecycle-node needs to be a nav2::Subscription

SteveMacenski commented on nav2_docking/opennav_docking/include/opennav_docking/simple_non_charging_dock.hpp:
```diff
@@ -127,6 +139,28 @@ class SimpleNonChargingDock : public opennav_docking_core::NonChargingDock
+  bool subscribe_toggle_{false};
+
+  // Client used to call the Trigger service
+  nav2_util::ServiceClient<
```
Ditto

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -196,6 +220,11 @@ bool SimpleChargingDock::getRefinedPose(geometry_msgs::msg::PoseStamped & pose,
   }
 
+  // Guard against using pose data before the first detection has arrived.
+  if (!detector_enabled_) {
```
Should log a warning a message

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +351,107 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+  return true;
+}
+
+void SimpleChargingDock::activate() {}
```
Remove override if this is not going to do anything

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +351,107 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+      });
+  }
+
+  RCLCPP_INFO(node_->get_logger(), "Detector START requested.");
```
Make this log a bit more formal :-)

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +351,107 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+  }
+
+  detector_enabled_ = false;
+  RCLCPP_INFO(node_->get_logger(), "Detector STOP requested");
```
Ditto

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -322,6 +351,107 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+      "detected_dock_pose", 1,
+      [this](const geometry_msgs::msg::PoseStamped::SharedPtr pose) {
+        detected_dock_pose_ = *pose;
+        detector_enabled_ = true;
```
Remove. This is the detector state not the topic state

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -105,6 +116,22 @@ void SimpleChargingDock::configure(
+      "detected_dock_pose", 1,
+      [this](const geometry_msgs::msg::PoseStamped::SharedPtr pose) {
+        detected_dock_pose_ = *pose;
+        detector_enabled_ = true;
```
Ditto

SteveMacenski reviewed (commented):
Overall, much improved! I still see a number of previous comments that were not fixed yet. Make sure to un-collapse the "see X more items" in case that's why they were missed. 

All feedback from the simple charging dock should also be applied to the non charging dock plugin as well since they're 1:1

bkoensgen commented:
Thanks for the detailed feedback, @SteveMacenski! It all makes sense. I'll get this rebased on main and updated. Will let you know when it's ready for another look

bkoensgen commented:
hi @SteveMacenski, ready for a fresh pass.

- Server control: `DockingServer` calls `startDetectionProcess()` before the `getRefinedPose()` loop, and `stopDetectionProcess()` on every terminal
path; order standardized after `publishZeroVelocity()` and before `succeeded_current()` terminate.
- APIs/migration: moved to `nav2::LifecycleNode`, `nav2::ServiceClient`, and `nav2::Subscription`; using `nav2::qos::StandardTopicQoS`.
- Subscriptions: `subscribe_toggle` defaults to false (code + docs). When toggling is disabled, the persistent subscription is created in configure().
- State/race: replaced the mini state machine with two flags `detection_started_` and `initial_pose_received_`; `getRefinedPose()` guards until the first pose arrives and checks freshness.
- Errors: start/stop return bool; start failure logs at ERROR and is propagated to the server.
- Docs: I’ll open the docs.nav2 PR (config + migration notes for `DockingServer` and new params) once you confirm the API/defaults here. If you prefer it upfront, I can send it now.

Open to confirm: `stopDetectionProcess()` placement (after `publishZeroVelocity(),` before `succeeded_current()`).

SteveMacenski reviewed (changes requested):
Closer! There are some still open items from my July 2 review that I added a "^" comment to for still relevant. Make sure to expand the comments if the GitHub UX collapses them due to the size of the thread in the PR.

But by in large, this looks very close to ready!

SteveMacenski commented:
@bkoensgen Checkout the test failure - its minor linting :-) https://app.circleci.com/pipelines/github/ros-navigation/navigation2/16074/workflows/42e6fc1f-9658-4d8b-b46d-0ba90539e3f2/jobs/46964/tests

SteveMacenski commented on nav2_docking/opennav_docking/src/docking_server.cpp:
```diff
@@ -167,6 +167,7 @@ DockingServer::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
   tf2_buffer_.reset();
   docking_action_server_.reset();
   undocking_action_server_.reset();
+  dock_db_->deactivate();
```
Remove, this should already be deactivated if we're in the cleanup transition

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -105,6 +116,26 @@ void SimpleChargingDock::configure(
+  node_->get_parameter(name + ".subscribe_toggle", subscribe_toggle_);
+
+  // Initialize detection state
+  detection_started_ = false;
```
Rename here and in the noncharging plugin to be `detection_active_` to be more clear

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -323,6 +356,118 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+void SimpleChargingDock::cleanup()
+{
+  detector_client_.reset();
+  dock_pose_sub_.reset();
```
Set 
```

  detection_started_ = false;
  initial_pose_received_ = false;
```

SteveMacenski commented on nav2_docking/opennav_docking/src/simple_charging_dock.cpp:
```diff
@@ -323,6 +356,118 @@ void SimpleChargingDock::jointStateCallback(const sensor_msgs::msg::JointState::
+
+  // 2. Subscription toggle
+  //    Only subscribe once; will set state to ON on first message
+  if (subscribe_toggle_ && !dock_pose_sub_) {
```
I'm not directly asking you to do this, just a point of discussion:

So in the startup of the sensor, we could, in concept, have a while loop to spin/block for the perception timeout, then return true/false if we obtain data or not. That would remove the `initial_pose_received_` variable since we wouldn't continue if we didn't obtain it.

SteveMacenski commented:
@bkoensgen can you handle the conflict? Please open a PR on docs.nav2.org migration guide to add the new feature + Configuration Guide for the new parameters added

Otherwise LGTM. I'll do a quick once over before merging but seems good to me if CI passes now.

bkoensgen commented:
So MergyBot should be happy now, and I’ve created the PR. Here’s the link: https://github.com/ros-navigation/docs.nav2.org/pull/773

mini-1235 commented:
This PR actually reverts some of the changes in https://github.com/ros-navigation/navigation2/pull/5521, I think we need to revert the changes related to `rclcpp::spin_some`
